StreamDeckPlugin/Services/DynamicActionInfoStore.cs

```diff
@@ -57,26 +60,73 @@ public class DynamicActionInfoStore : IDynamicActionInfoStore {
-            var changedActionInfoList = new List<DynamicActionInfo>();
+            DynamicActionInfo changedAction;
             lock (_cacheLock) {
                 var itemExistsAtLocation = _dynamicActionInfoList.FirstOrDefaultWithContext(buttonContex) != null;
                 if (itemExistsAtLocation) {
                     //make room for the item we are inserting by shifting the index of everything else up
                     foreach (var dynamicActioninfo in _dynamicActionInfoList) {
                         if (dynamicActioninfo.IsAtSameIndexOrAfter(buttonContex)) {
                             dynamicActioninfo.Index++;
-                            changedActionInfoList.Add(dynamicActioninfo);
                         }
                     }
                 }
 
-                var dynamicActionInfo = new DynamicActionInfo(buttonContex);
-                dynamicActionInfo.UpdateFromCardInfo(cardInfo);
-                _dynamicActionInfoList.Add(dynamicActionInfo);
-                changedActionInfoList.Add(dynamicActionInfo);
+                changedAction = new DynamicActionInfo(buttonContex);
+                changedAction.UpdateFromCardInfo(cardInfo);
+                _dynamicActionInfoList.Add(changedAction);
             }
 
-            PublishChangeEventsForChangedActions(changedActionInfoList);
+            //we only have to change the one action that changed, as the handler is smart enough to refresh everything else around
+            PublishChangeEventsForChangedActions(changedAction);
+        }
+
+        public IEnumerable<IDynamicActionInfo> GetDynamicActionInfoForGroup(CardGroupId cardGroupId) {
+            lock (_cacheLock) {
+                return (from dynamicActionInfo in _dynamicActionInfoList
+                        where dynamicActionInfo.CardGroupId == cardGroupId
+                        select dynamicActionInfo).ToList();
+            }
+        }
+
+        /// <summary>
+        /// Clear the current list of buttons for this card group and replace it with a new list
+        /// </summary>
+        /// <param name="eventData">Event data containing list of new buttons</param>
+        private void CardGroupButtonsChangedHandler(CardGroupButtonsChanged eventData) {
+            IList<DynamicActionInfo> actionsUpdated = new List<DynamicActionInfo>();
+
+            lock (_cacheLock) {
+                var removedActions = (from dynamicActionInfo in _dynamicActionInfoList
+                                  where dynamicActionInfo.CardGroupId == eventData.CardGroupId
+                                  select dynamicActionInfo).ToList();
+
+                foreach (var actionToRemove in removedActions) {
+                    _dynamicActionInfoList.Remove(actionToRemove);
+                }
+
+                foreach (var button in eventData.Buttons) {
+                    var changedAction = new DynamicActionInfo(button);
+                    changedAction.UpdateFromCardInfo(button);
```

Comment:
Minor: This feels unintuitive to me, since we pass in the same object twice, Consider adding a constructor that does both these lines.

Reply:
I'm not sure how to do this without something that looks like new DynamicActionInfo(button, button), which also looks unintuitive. The issue is that the button is implementing two different interfaces- one that identifies the button, and one that contains information about the button.

Reply:
That's fair. I think this is fine, because on closer inspection you can see why it's doing this and the `UpdateFromCardInfo` has other uses.